ouroboros-consensus/src/Ouroboros/Consensus/Protocol/Abstract.hs

```diff
@@ -90,6 +90,44 @@ class ( Show (ChainState    p)
   type family IsLeader p :: *
 
   -- | Projection of the ledger state the Ouroboros protocol needs access to
+  --
+  -- The 'LedgerView' is a summary of the state of the ledger that the consensus
+  -- algorithm requires to do its job. Under certain circumstances the consensus
+  -- algorithm may require the 'LedgerView' for slots in the past (before the
+  -- current tip of the chain) or in the (near) future (beyond the tip of the
+  -- current chain, without having seen those future blocks yet).
+  --
+  -- This puts limitations on what the 'LedgerView' can be. For example, it
+  -- cannot be the "current stake distribution", since it is of course
+  -- impossible to compute the current stake distibution for a slot in the
+  -- future. This means that for a consensus algorithm that requires the
+  -- stake distribution such as Praos, the 'LedgerView' for a particular slot
+  -- must be the "stake distribution for the purpose of leader selection".
+  -- This "active" stake distribution /can/ be computed for slots in the
+  -- (near) future because it is based on historical stake, not current.
+  --
+  -- A somewhat unfortunate consequence of this is that some decisions that
+  -- ought to live in the consensus layer (such as the decision precisely which
+  -- historical stake to sample to determine the active stake distribution)
+  -- instead live in the ledger layer. It is difficult to disentangle this,
```

Comment:
I'm not completely convinced this is really necessary.

Say we're interested in the stake distribution that determined the leader schedule for the epoch of slot `s`, and that that stake distribution is the one from slot `s'`. In your proposal, the consensus layer would ask the ledger layer for the "active stake distribution" for slot `s`, and the ledger layer would have to compute `s'` and return the stake distribution from that slot. This will work, as long as `s'` is not in the future.

Could we not just as well let the consensus layer figure out what `s'` is, and then ask the ledger layer for the stake distribution at `s'`? The only time that will not work is when `s'` itself is in the future, and we won't have an answer then in any case.

Am I missing something?

It might still be the best thing to let the ledger layer decide how to compute the "active stake distribution", but I'm not convinced we are forced to implement it that way.

Reply:
@kantp I think the problem is that reward calculation depends on that _same_ decision. It would not be correct for the consensus layer to sample at slot `s'` and the reward calculation to sample at slot `s''`, because rewards wouldn't make much sense anymore if they are based on a different leader schedule. But I've love to be wrong about that :)

Reply:
Sorry, but you're not wrong :)

So yes, we either need the ledger layer to determine `s'`, or have some way for the consensus layer to tell the ledger layer about `s'`, which sounds pretty convoluted.